Source/WebCore/css/StyleProperties.cpp

```diff
@@ -576,12 +577,34 @@ String StyleProperties::fontValue() const
     if (auto shorthand = isSingleFontShorthand())
         return getValueNameAtomString(shorthand.value());
 
+    // Font stretch values can only be serialized as keywords.
+    // While it's not clear the CSS font specification calls for this, web platform tests expect us to convert percentages to keywords.
```

Comment:
The CSS font spec says percentages for `font-stretch` should get rejected from the `font` shorthand, so that translation is necessary in order to retain a correct value. https://drafts.csswg.org/css-fonts-4/#font-prop . So before this patch, we were effectively serializing to a non valid value.

It is in accordance of step 2 of the CSSOM spec too: https://drafts.csswg.org/cssom/#serializing-css-values

I'd just remove that specific line of the comment re. the spec.

Reply:
Sure, it’s obvious that the old serialization was incorrect. But we could serialize to an empty string in that case. Actually _translating_ percentages to keywords is unusual and unprecedented. And this is what the test expects and requires.

Reply:
I studied the CSSOM specification and I saw **nothing** about converting one type of value (a percentage) into another (a keyword). I’m not sure what you are referring to when you say it’s there. There’s stuff about using a shorter form, and only serializing things that will round trip, but turning a number into a keyword would not be implied by what is written there and i think this is an example of implementers going beyond what is specified.

Reply:
Actually, I would say it's implied by step 1:

1. If this algorithm is invoked with a [list](https://infra.spec.whatwg.org/#list) list:
  1. Let shorthand be the first shorthand property, in [preferred order](https://drafts.csswg.org/cssom/#concept-shorthands-preferred-order), that exactly maps to all of the longhand properties in list.
  2. If there is no such shorthand or shorthand cannot exactly represent the values of all the properties in list, return the empty string.
  3. Otherwise, [serialize a CSS value](https://drafts.csswg.org/cssom/#serialize-a-css-value) from a hypothetical declaration of the property shorthand with its value representing the combined values of the declarations in list.

It would go all the way through step 1.3, and given percentages cannot be represented as they are in the shorthand, they have to be translated. Condition 1.2 is not met here, since the shorthand can exactly represent all the values with some translation.

(I think I mentioned 2. because that step mentions backwards compatibility, but I realize 1. is more accurate)

Reply:
OK, I’ll stop arguing this point with you, and remove the comment. But for the record, please note that step 1.3 says "return the empty string". It does not mention translating anything.

Reply:
It all depends if you think "represent the values" means matching the computed values or matching the specified values. A specified keyword is not the same as a specified percentage in CSS even if they result in the same computed value.

Reply:
I don’t know if this is relevant, but this is the opposite behavior from font-weight. I had to write a patch to stop converting from numbers to keywords for font-weight.

Reply:
For now I will land this code, and I guess I don’t need to put that comment in, although for clarity I would like to. The discussion of whether this behavior that Firefox and Chrome did and put into the WPT test is correct or not should probably be a _specification_ discussion, not a WebKit bug report discussion. One good thing about WPT is that if we do change our minds we won’t forget to change the code — it will be a test failure!